### ISSUE TITLE
Prometheus returning zero for some queries instead of NaN

### ISSUE THREAD

**Issue by bohandley** [grafana/grafana#54857]
It was reported that between v8.5.0 and v9.2.0 that Prometheus was returning zero instead of NaN. There was a breaking change in v9.0.0 that Prometheus would only return NaN but for some reason, zero is being returned in the following query.

```
sum(avg_over_time(cluster_namespace_service_slo:successful_requests_total:rate5m{cluster="ops-us-east-0", namespace=~".+", service="cortex", slo="cardinality"}[$__rate_interval]))
/
sum(avg_over_time(cluster_namespace_service_slo:requests_total:rate5m{cluster="ops-us-east-0", namespace=~".+", service="cortex", slo="cardinality"}[$__rate_interval]))
```

Original thread: https://raintank-corp.slack.com/archives/CPXKS6FB5/p1662559567729429

[v8.5.0](https://admin-ops-us-east-0.grafana.net/stable-grafana/explore?orgId=1&left=%7B%22datasource%22:%22prometheus%22,%22queries%22:%5B%7B%22exemplar%22:true,%22refId%22:%22A%22,%22expr%22:%22sum(avg_over_time(cluster_namespace_service_slo:successful_requests_total:rate5m%7Bcluster%3D%5C%22ops-us-east-0%5C%22,%20namespace%3D~%5C%22.%2B%5C%22,%20service%3D%5C%22cortex%5C%22,%20slo%3D%5C%22cardinality%5C%22%7D%5B$__rate_interval%5D))%5Cn%2F%5Cnsum(avg_over_time(cluster_namespace_service_slo:requests_total:rate5m%7Bcluster%3D%5C%22ops-us-east-0%5C%22,%20namespace%3D~%5C%22.%2B%5C%22,%20service%3D%5C%22cortex%5C%22,%20slo%3D%5C%22cardinality%5C%22%7D%5B$__rate_interval%5D))%22,%22hide%22:false,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true,%22interval%22:%225m%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)
![image](https://user-images.githubusercontent.com/25674746/188920338-c9b7354e-2b7b-4baf-8b54-03617c13fd99.png)

[v9.2.0](https://admin-ops-us-east-0.grafana.net/grafana/explore?orgId=1&left=%7B%22datasource%22:%22P1809F7CD0C75ACF3%22,%22queries%22:%5B%7B%22exemplar%22:true,%22refId%22:%22A%22,%22expr%22:%22sum%28avg_over_time%28cluster_namespace_service_slo:successful_requests_total:rate5m%7Bcluster%3D%5C%22ops-us-east-0%5C%22,%20namespace%3D~%5C%22.%2B%5C%22,%20service%3D%5C%22cortex%5C%22,%20slo%3D%5C%22cardinality%5C%22%7D%5B$__rate_interval%5D%29%29%5Cn%2F%5Cnsum%28avg_over_time%28cluster_namespace_service_slo:requests_total:rate5m%7Bcluster%3D%5C%22ops-us-east-0%5C%22,%20namespace%3D~%5C%22.%2B%5C%22,%20service%3D%5C%22cortex%5C%22,%20slo%3D%5C%22cardinality%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22hide%22:false,%22editorMode%22:%22code%22,%22range%22:true,%22instant%22:true,%22interval%22:%225m%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P1809F7CD0C75ACF3%22%7D%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)
![image](https://user-images.githubusercontent.com/25674746/188920408-57ba4011-3d0b-4dd4-8135-c6cab98d12d6.png)

**Comment by gtk-grafana** [comment]
Trying to take a look at this in between other tasks.

**Comment by gtk-grafana** [comment]
This was introduced in 9.1.0-beta with https://github.com/grafana/grafana/pull/51711. Here's a PR https://github.com/grafana/grafana/pull/55052 reverting this change which makes 0/0 return the same null values as 8.5.0

**Comment by gtk-grafana** [comment]
@aocenas Is this actually a bug, or a feature? It looks like the fix is trivial, but I don't have the context to know if this is something we want to fix or not!

**Comment by kylebrandt** [comment]
Both float and nullable float fields support NaN.

I think (haven't made tests) it comes from the following lines (where we don't set the value if it is NaN):

https://github.com/grafana/grafana/blob/14d4e0eeca72cb4b54eb5091fe8c8c100dfa7ec8/pkg/tsdb/prometheus/buffered/time_series_query.go#L371-L373

So I don't think changing it to nullable will change this behavior.

These lines come from https://github.com/grafana/grafana/pull/44380 . It looks like that left off with corridating a possible breaking change with alerting since the SSE feature to drop NaNs etc exists now?

**Comment by gtk-grafana** [comment]
@kylebrandt When I make this change locally, it does appear to fix the bug. Did you try running that PR locally? Also Andrej weighed in on the original PR: https://github.com/grafana/grafana/pull/51711#discussion_r969345882

**Comment by kylebrandt** [comment]
So switching to nullable would result return null if prom returns NaN I expect. That doesn't seem like correct behavior to me?

**Comment by gtk-grafana** [comment]
Yeah it seems weird to me too, but the requested change would bring the behavior back in line with how 8.5 operates. Also we'll want to address making this change to the streaming parser as well if we decide that this is the desired functionality.

**Comment by kylebrandt** [comment]
So if we go changing this back to nullable, and therefore null values (or a change to NaN values), it may change the behavior of Grafana Managed Alerts.

For example, currently in main (non-nullable, and has `!math.IsNaN` condition):

![image](https://user-images.githubusercontent.com/1692624/189913535-5454ef17-69d2-41ec-a8a0-8d3a7964bb7c.png)

But changing this to nullable and keep the `!math.IsNaN` condition, the behavior changes to alerting:

![image](https://user-images.githubusercontent.com/1692624/189913683-20cd3519-fe1b-4e79-804b-bc0eb0caa824.png)

This can be worked around by using the newer "replace" option in the SSE Reduce Math operation, but users may not have configured this. So I expected the break change needs to be highlighted in a release, and we need an okay from alerting:

![image](https://user-images.githubusercontent.com/1692624/189914049-2cbc48c6-ca7a-497d-86d2-e133a40fb141.png)

But since the `math.IsNaN` condition was added for alerting, I think we are likely better off passing what prometheus returns and not changing the values in the data source (which is NaN) if we are going for a change -- (Although need to see how this impacts visualizations)

**Comment by gtk-grafana** [comment]
@stevesg (copying [question](https://raintank-corp.slack.com/archives/CPXKS6FB5/p1663344126871009?thread_ts=1662559567.729429&cid=CPXKS6FB5) in slack here)

We’re kinda under the impression that sending null isn’t the ideal functionality, and we’d expect to see NaN instead of mapping this to null.
Should we be looking into fixing this to return what prometheus is returning (NaN), or do we just want to fix the regression so modern grafana returns the same result as stable (i.e. null)?

**Comment by stevesg** [comment]
For the specific case I was looking at, I don't have a strong opinion about seeing `NaN`/`Inf` vs `null`, as long as they are **_not_** converted to 0 - that feels wrong.

**Comment by gtk-grafana** [comment]
Can someone from @grafana/alerting-squad-backend weigh in? It looks like we want to start returning NaN, but we're concerned that might be a breaking change.
Also if we want to return NaN to the clients instead of null/0, will we need to take extra care when decoding the JSON since NaN is not part of JSON?

**Comment by joeblubaugh** [comment]
The alerting backend treats `nil` and NaN differently. Once an expression is evaluated, `nil` is treated as "NoData" and `NaN` is treated as a non-zero value. [Non-zero values generate alert instances. ](https://github.com/grafana/grafana/blob/4df41972f11456537994320f2ae39124ecf55fea/pkg/services/ngalert/eval/eval.go#L380)

So changing this behaviour in 9.0 changed the output of alert evaluation from "NoData" to "not alerting". The outcome of that depends on the alert rule's configuration for "NoData" but it could have reduced the number of alerts delivered.

Changing this to NaN *will always* generate alert states because NaN is a non-zero value.

**Comment by joeblubaugh** [comment]
Returning NaN will generate significantly more alerts than returning nil. I think that `nil` is probably the better approach.

**Comment by gtk-grafana** [comment]
@kylebrandt Looks like we'll want to revert this back to how it was behaving to return nil/null instead of 0 or NaN. I think the PR for that was just a simple 1 line fix, I'll try to dig that up and get the PR opened for this.

**Comment by gtk-grafana** [comment]
Here it is: https://github.com/grafana/grafana/pull/55052

**Comment by kylebrandt** [comment]
We should restore the behavior to returns NaNs (matching what Prometheus returns).

As a general rule we do not want avoid mutating data returned from the data store api behind a datasource as much as possible:
* It makes for a more confusing user experience if what you get in Grafana doesn't match the behavior of the datastore behind the datasource itself. "Raw Views" won't match the documentation of the backing datastore.
* Since multiple consumers use this data (Explore Viz, Recorded Queries, SSE, Grafana Managed Alerts) we get unintended consequences when we change the datasource to fix a consumer. So sticking to "don't mutate data" gives us a longer term development foundation to avoid conflict between the needs of consumers.

In this specific case of Prometheus we also want not use a nullable float for performance reasons. Since a `SampleValue` in prom appears to not be nullable, we want to avoid the performance hit of pointers to insert nulls when there are no nulls from the source (as far as I can tell). This is part of streaming parser PR.

So we should move to NaN. We have been going on circles on this a bit since [2019](https://github.com/grafana/grafana/pull/18284#issuecomment-515976813), so have this decision fixed in place, so things upstream can adjust accordingly.

I'm not sure of how all consumers will be impacted. Alerting users may have to use the new Reduce SSE replace options for non-numeric values, and visualization may change what it does with connected lines? But this is the best long term way forward to consistent logical behavior across the platform with the datasource.

I think the change will come in with the move to the [prom json stream parser](https://github.com/grafana/grafana/pull/55941), so we need to capture this in the changelog.

**Comment by joeblubaugh** [comment]
@grafana/alerting-squad @armandgrillet going with NaN for this change is going to require a flashy announcement in the release notes, because this may be a big behavior change for existing alerting users.

**Comment by gtk-grafana** [comment]
@kylebrandt is this issue closed?

**Comment by candlerb** [comment]
I think I am seeing the same issue here, although in this case it's not a PromQL expression dividing two values which returns a NaN, but an actual NaN stored in the timeseries.

I was expecting NaN to render as a gap in the graph (like a missing data point), but since the data source converts NaN to zero, I get a line drawn from the zero axis.

![image](https://user-images.githubusercontent.com/44789/203072234-7fcaf97a-565c-410a-b2c8-59d27e48b79f.png)

I can demonstrate that this is definitely Grafana's data source mapping NaN to zero.

**How to reproduce it (as minimally and precisely as possible)**:

I have an exporter which returns a "NaN" value for a short time after it has been restarted, before giving the true value.  Here are the raw query results from directly querying the Prometheus API (set to the same time interval and step as the graph shown):

```
# /opt/prometheus/promtool query range --start=1669035217.294 --end=1669035921.515 --step=30s http://127.0.0.1:9090/prometheus 'solis_battery_soc{instance="solis:3105"}'
solis_battery_soc{instance="solis:3105", job="solis"} =>
65 @[1669035217.294]
65 @[1669035247.294]
65 @[1669035277.294]
65 @[1669035307.294]
65 @[1669035337.294]
65 @[1669035367.294]
65 @[1669035397.294]
65 @[1669035427.294]
64 @[1669035457.294]
64 @[1669035487.294]
NaN @[1669035607.294]
64 @[1669035637.294]
64 @[1669035667.294]
64 @[1669035697.294]
64 @[1669035727.294]
63 @[1669035757.294]
63 @[1669035787.294]
62 @[1669035817.294]
62 @[1669035847.294]
62 @[1669035877.294]
62 @[1669035907.294]
```

And here is the corresponding captured Grafana query from Inspect > Query > Refresh:

<details>
<summary>Click to expand</summary>

```
{
  "request": {
    "url": "api/ds/query",
    "method": "POST",
    "data": {
      "queries": [
        {
          "datasource": {
            "uid": "000000001",
            "type": "prometheus"
          },
          "editorMode": "code",
          "exemplar": true,
          "expr": "solis_battery_soc{instance=\"solis:3105\"}",
          "interval": "",
          "legendFormat": "solis_battery_soc",
          "range": true,
          "refId": "A",
          "queryType": "timeSeriesQuery",
          "requestId": "18A",
          "utcOffsetSec": 0,
          "datasourceId": 1,
          "intervalMs": 30000,
          "maxDataPoints": 396
        }
      ],
      "range": {
        "from": "2022-11-21T12:53:37.294Z",
        "to": "2022-11-21T13:05:21.515Z",
        "raw": {
          "from": "2022-11-21T12:53:37.294Z",
          "to": "2022-11-21T13:05:21.515Z"
        }
      },
      "from": "1669035217294",
      "to": "1669035921515"
    },
    "hideFromInspector": false
  },
  "response": {
    "results": {
      "A": {
        "frames": [
          {
            "schema": {
              "name": "solis_battery_soc",
              "refId": "A",
              "meta": {
                "type": "timeseries-many",
                "custom": {
                  "resultType": "matrix"
                },
                "executedQueryString": "Expr: solis_battery_soc{instance=\"solis:3105\"}\nStep: 30s"
              },
              "fields": [
                {
                  "name": "Time",
                  "type": "time",
                  "typeInfo": {
                    "frame": "time.Time"
                  },
                  "config": {
                    "interval": 30000
                  }
                },
                {
                  "name": "Value",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64"
                  },
                  "labels": {
                    "__name__": "solis_battery_soc",
                    "instance": "solis:3105",
                    "job": "solis"
                  },
                  "config": {
                    "displayNameFromDS": "solis_battery_soc"
                  }
                }
              ]
            },
            "data": {
              "values": [
                [
                  1669035210000,
                  1669035240000,
                  1669035270000,
                  1669035300000,
                  1669035330000,
                  1669035360000,
                  1669035390000,
                  1669035420000,
                  1669035450000,
                  1669035480000,
                  1669035510000,
                  1669035600000,
                  1669035630000,
                  1669035660000,
                  1669035690000,
                  1669035720000,
                  1669035750000,
                  1669035780000,
                  1669035810000,
                  1669035840000,
                  1669035870000,
                  1669035900000
                ],
                [
                  65,
                  65,
                  65,
                  65,
                  65,
                  65,
                  65,
                  65,
                  64,
                  64,
                  64,
                  0,
                  64,
                  64,
                  64,
                  64,
                  63,
                  63,
                  62,
                  62,
                  62,
                  62
                ]
              ]
            }
          },
          {
            "schema": {
              "name": "exemplar",
              "refId": "A",
              "meta": {
                "custom": {
                  "resultType": "exemplar"
                },
                "executedQueryString": "Expr: solis_battery_soc{instance=\"solis:3105\"}\nStep: 30s"
              },
              "fields": [
                {
                  "name": "Time",
                  "type": "time",
                  "typeInfo": {
                    "frame": "time.Time"
                  }
                },
                {
                  "name": "Value",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64"
                  }
                }
              ]
            },
            "data": {
              "values": [
                [],
                []
              ]
            }
          }
        ],
        "refId": "A"
      }
    }
  }
}
```
</details>

The key part to note is the values array:

```
                  64,
                  64,
                  0,
                  64,
                  64,
```

Grafana's datasource has mapped NaN to zero, and therefore it gets rendered as a point on the zero axis.

**Environment**:
- Grafana version: 9.2.5
- Data source type & version: Prometheus 2.37.1
- OS Grafana is installed on: Ubuntu 22.04.5
- User OS & Browser: macOS, Chrome
- Grafana plugins: none
- Others: n/a

**Comment by pragkent** [comment]
Still seeing this in 9.2.6, is this issue fixed?

**Comment by kylebrandt** [comment]
This should be fixed in v9.3

We changed the datasource to use the new streaming parser client in the prometheus datasource (more performant) by default, and tied the change of preserving NaNs to this new client.
